### ISSUE TITLE
Allow to set `files.encoding` as language specific setting for files on startup

### ISSUE THREAD

**Issue by VisionXi** [microsoft/vscode#127936]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.57.1, but also previous versions
- OS Version: Windows 10 and 7, all versions until 2021, July

Well know bug:

Language specific resource settings are not resolved for files opening right on startup · Issue #39084 and others.

Please reopen this issue and set priority on it. It's really worse and there are many people running into this issue!

Perhaps you find a way to handle this by a configurable seeting or command line switch.

**Comment by CodeCrazy-ywt** [comment]
@vscode-triage-bot I sincerely suggest you to look into this bothering issue that a large number of users strongly require to be solved. Many thanks.

**Comment by VisionXi** [comment]
Sorry, I don't know how to get priority on this issue. It seems that the bot has closed this issue.  But it's really annoying and such an important use case.
If it's really because the extension host has to scan all extensions and for performance reasons we do not want to wait on that, then why not trying to solve this by a configurable setting. I could live with a small delay when opening a file with vscode. This delay would only occur if vscode was not opened before.

**Comment by bpasero** [comment]
Renamed to reflect the issue: my understanding is that this only applies to settings which we need right in the beginning when opening a file, specifically: the encoding and the line endings. All other settings should apply after the extensions have been resolved.

I briefly looked into this and still don't have a good solution, other than maybe caching all the languages mapped to file types to have them ready right in the beginning.

We resolve the language of the file here:

https://github.com/microsoft/vscode/blob/58e840c3738db005e010d58a4294a3eb2667f0b4/src/vs/editor/common/services/textResourceConfigurationServiceImpl.ts#L114-L114

This code will eventually hit `guessMimeTypes`:

https://github.com/microsoft/vscode/blob/b776d802d999fd769fd884c15e0c6ecac933ee0b/src/vs/base/common/mime.ts#L114-L114

But in order for this to work, we need `registerTextMime` to be called with all languages, including extensions:

https://github.com/microsoft/vscode/blob/b776d802d999fd769fd884c15e0c6ecac933ee0b/src/vs/base/common/mime.ts#L44-L44

Which is not the case for files opened right on startup.

//cc @sandy081 @alexdima for more ideas

**Comment by bpasero** [comment]
Maybe we could have new API on the `IConfigurationService` that I can ask whether any language specific configuration exists for `files.encoding` or `files.eol` and only if so, prolong the opening of a file on startup until languages have been registered.

However, I just checked our default settings and it looks like we actually define `files.eol` for shell scripts:

```json
"[shellscript]":  {
   "files.eol": "\n"
}
```

So we need to be a bit more clever here, otherwise we would always wait for languages being registered from extensions...

**Comment by sandy081** [comment]
`inspect` API provides you the override identifiers

https://github.com/microsoft/vscode/blob/bdd429183e983a0a9bb953951a33b9e14e62331c/src/vs/platform/configuration/common/configuration.ts#L115-L115

https://github.com/microsoft/vscode/blob/bdd429183e983a0a9bb953951a33b9e14e62331c/src/vs/platform/configuration/common/configuration.ts#L68-L88

**Comment by gjsjohnmurray** [comment]
Maybe allow file-extension-specific overrides in addition to language-specific ones.

**Comment by VisionXi** [comment]
Is there any progress in solving this issue? I understand that it's not easy. But it's really annoying if you want to open a file direclty from a file manager and not within VC.

**Comment by bpasero** [comment]
Not super happy with the solution I pushed now and it might backfire, but here is what is happening now:
* we now detect changes to the editor language and there is always a transition from `plaintext` in the beginning until languages are resolved
* if we detect that the new language has a configured `files.encoding` override
* and that encoding is different from the current encoding
* and the editor is not dirty or in save conflict resolution
* we reopen the file with the configured encoding

Unfortunately I cannot tell apart this from happening vs. the user changing the language from the editor status bar. So if the user changes language mode to `bat` from to `ps1` and `ps1` has a configured encoding, the encoding will change. 

🤷

**Comment by Raydir** [comment]
sorry, that i did not write soon enough in #39084 soon enough, but for me this does not work the way it "should" to be honestly...

since i opened #39084 we _partly_  changed some of our scripts to UTF8bom.

So while migrating our customers to the latest version of our software (what could take some time...) we are in some kind of mixedmode.

My Extension does set the files.encoding param to OEM850 (due to old files). if there is explicity an file that does deliver a UTF8 with a BOM this should not be overridden by the files.encoding param.

UTF8bom does explicit deliver the BOM so that e.g. Editors can open the file correctly.

-> "files.autoGuessEncoding": true does not work becauses this is setting the OEM850 files to OEM866 (...because that is the first implemented encoding in the OEM sector i gues...)

Thanks for a second insight on this matter.

**Comment by bpasero** [comment]
@Raydir can you open a new issue with some basic steps how to reproduce, I am not sure I understand the issue or if its related to this issue.
